ChurchCRM/CRM
# Feature/dashboard service

## Note that, when switching to this branch you have to run:
```
cd src
composer dump-autoload
```
to update the namespace autoloader

#### What's this PR do?
Moves handling dashboard-style elements into individual PHP-delcared DTOs.

Watch dashboard values change without reloading the page as data is entered from another session:
@phili67  your event counters work ;-)

(Animated gif):
![livedashboard](https://user-images.githubusercontent.com/11679900/33590304-61758904-d94c-11e7-8325-b8c869d2038c.gif)


This is all done by a single recurring AJAX call that will contain different data based on the current page.  For example - ```/api/dashboard/page/Menu.php```:

![image](https://user-images.githubusercontent.com/11679900/33590379-e3140df0-d94c-11e7-9d2e-1a79ec8a2447.png)


#### What Issues does it Close?

Closes #3612
Closes #3643 
Closes #3640 

Supersedes #3645 

#### Where should the reviewer start?

1.  Examine the new *DashboardItem.php classes at src\ChurchCRM\Dashboard.   These are meant to encapsulate the query and presentation logic for rendering any "live-updating" dashboard widgets.

2.  Look at the new Dashboard.js file - this will call the new dashboard API endpoint on a regular basis, and execute the function stored in ```window.CRM.dashboard.{dashboard widget name}``` 

3. Look at Footer.php for the include of the auto-generated client-side code to manage updating the UI.  This code comes from the individual DashboardItem classes.   

4.  Note that all API endpoints reference the "current page"  so that only the dashboard items which respond ```true``` when the page's name satisifies the ```shouldInclude``` function 

#### How should this be manually tested?
1.  Open a web browser to the main dashboard.  
2.  Watch the network and Javascript console - you should see the dashboard updsates
3.  Using a separate browser, alter some data that would trigger a change in the dashboard elements visible in the first browser.  The values should change without a page reload

#### How should the automated tests treat this?
No.

#### Any background context you want to provide?


#### What are the relevant tickets?


#### Screenshots (if appropriate)
